File: day13/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"sort"
)

const (
	puzzleInput = "input.txt"
)

const (
	dirN = 0
	dirE = 1
	dirS = 2
	dirW = 3
)

type (
	Cart struct {
		x, y, dir int
		turn      int
	}

	CartList struct {
		carts CartSlice
		board [][]byte
	}

	CartSlice []*Cart
)

func (s CartSlice) Len() int {
	return len(s)
}
func (s CartSlice) Swap(i, j int) {
	s[i], s[j] = s[j], s[i]
}
func (s CartSlice) Less(i, j int) bool {
	if s[i].y == s[j].y {
		return s[i].x < s[j].x
	}
	return s[i].y < s[j].y
}

func cartCharToDir(b byte) int {
	switch b {
	case byte('^'):
		return dirN
	case byte('>'):
		return dirE
	case byte('v'):
		return dirS
	case byte('<'):
		return dirW
	default:
		return -1
	}
}

func cartDirToChar(dir int) byte {
	switch dir {
	case dirN:
		return byte('^')
	case dirE:
		return byte('>')
	case dirS:
		return byte('v')
	case dirW:
		return byte('<')
	default:
		panic("invalid dir")
	}
	return 8
}

func NewCart(x, y, dir int) *Cart {
	return &Cart{
		x:    x,
		y:    y,
		dir:  dir,
		turn: 0,
	}
}

func (c *Cart) Step(board [][]byte) {
	switch c.dir {
	case dirN:
		c.y--
	case dirE:
		c.x++
	case dirS:
		c.y++
	case dirW:
		c.x--
	}
	switch board[c.y][c.x] {
	case byte('\\'):
		switch c.dir {
		case dirN:
			c.dir = dirW
		case dirE:
			c.dir = dirS
		case dirS:
			c.dir = dirE
		case dirW:
			c.dir = dirN
		default:
			panic("invalid direction")
		}
	case byte('/'):
		switch c.dir {
		case dirN:
			c.dir = dirE
		case dirE:
			c.dir = dirN
		case dirS:
			c.dir = dirW
		case dirW:
			c.dir = dirS
		default:
			panic("invalid direction")
		}
	case byte('+'):
		switch c.turn {
		case 0:
			c.dir = (c.dir + 3) % 4
		case 1:
		case 2:
			c.dir = (c.dir + 1) % 4
		default:
			panic("invalid turn")
		}
		c.turn = (c.turn + 1) % 3
	}
}

func NewCartList(board [][]byte) *CartList {
	return &CartList{
		carts: CartSlice{},
		board: board,
	}
}

func (cl *CartList) AddCart(c *Cart) {
	cl.carts = append(cl.carts, c)
}

func (cl *CartList) Step() {
	crashed := map[int]struct{}{}
	for n, i := range cl.carts {
		if _, ok := crashed[n]; ok {
			continue
		}
		i.Step(cl.board)
		if crash := cl.HasCrash(i, crashed); crash > -1 {
			fmt.Println(i.x, i.y)
			crashed[n] = struct{}{}
			crashed[crash] = struct{}{}
		}
	}
	k := CartSlice{}
	for n, i := range cl.carts {
		if _, ok := crashed[n]; !ok {
			k = append(k, i)
		}
	}
	cl.carts = k
	sort.Sort(cl.carts)
}

func (cl *CartList) HasCrash(c *Cart, crashed map[int]struct{}) int {
	for n, i := range cl.carts {
		if _, ok := crashed[n]; ok {
			continue
		}
		if c != i && c.x == i.x && c.y == i.y {
			return n
		}
	}
	return -1
}

func (cl *CartList) Print() {
	board := make([][]byte, len(cl.board))
	for i := range board {
		board[i] = make([]byte, len(cl.board[i]))
		copy(board[i], cl.board[i])
	}
	for _, i := range cl.carts {
		if cartCharToDir(board[i.y][i.x]) > -1 {
			board[i.y][i.x] = byte('X')
		} else {
			board[i.y][i.x] = cartDirToChar(i.dir)
		}
	}
	for _, i := range board {
		fmt.Println(string(i))
	}
}

func (cl *CartList) PrintPos() {
	for _, i := range cl.carts {
		fmt.Printf("%d, %d; ", i.x, i.y)
	}
	fmt.Println()
}

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	board := [][]byte{}
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		board = append(board, []byte(scanner.Text()))
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	carts := NewCartList(board)
	for y, row := range board {
		for x, i := range row {
			if dir := cartCharToDir(i); dir > -1 {
				carts.AddCart(NewCart(x, y, dir))
				if dir == dirE || dir == dirW {
					board[y][x] = byte('-')
				} else if dir == dirN || dir == dirS {
					board[y][x] = byte('|')
				}
			}
		}
	}

	for i := 0; carts.carts.Len() > 1; i++ {
		carts.Step()
	}

	lastCart := carts.carts[0]
	fmt.Println(lastCart.x, lastCart.y)
}

File: day13/src/main.rs
use std::cmp::Ordering;
use std::collections::HashSet;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

enum Turn {
    Left,
    Straight,
    Right,
}

impl Turn {
    fn next(&self) -> Self {
        match self {
            Turn::Left => Turn::Straight,
            Turn::Straight => Turn::Right,
            Turn::Right => Turn::Left,
        }
    }
}

#[derive(Clone, Copy)]
enum Dir {
    Up,
    Right,
    Down,
    Left,
}

impl Dir {
    fn right(&self) -> Self {
        match self {
            Dir::Up => Dir::Right,
            Dir::Right => Dir::Down,
            Dir::Down => Dir::Left,
            Dir::Left => Dir::Up,
        }
    }

    fn left(&self) -> Self {
        match self {
            Dir::Up => Dir::Left,
            Dir::Right => Dir::Up,
            Dir::Down => Dir::Right,
            Dir::Left => Dir::Down,
        }
    }

    fn turn(&self, turn: &Turn) -> Self {
        match turn {
            Turn::Left => self.left(),
            Turn::Straight => *self,
            Turn::Right => self.right(),
        }
    }

    fn cart_char_to_dir(c: char) -> Option<Self> {
        match c {
            '^' => Some(Dir::Up),
            '>' => Some(Dir::Right),
            'v' => Some(Dir::Down),
            '<' => Some(Dir::Left),
            _ => None,
        }
    }
}

struct Cart {
    x: usize,
    y: usize,
    dir: Dir,
    turn: Turn,
}

impl Cart {
    fn new(x: usize, y: usize, dir: Dir) -> Self {
        Self {
            x: x,
            y: y,
            dir: dir,
            turn: Turn::Left,
        }
    }

    fn next_step(&self) -> (usize, usize) {
        match self.dir {
            Dir::Up => (self.x, self.y - 1),
            Dir::Down => (self.x, self.y + 1),
            Dir::Left => (self.x - 1, self.y),
            Dir::Right => (self.x + 1, self.y),
        }
    }

    fn tick(&mut self, board: &Vec<Vec<char>>) {
        match self.dir {
            Dir::Up => self.y -= 1,
            Dir::Down => self.y += 1,
            Dir::Left => self.x -= 1,
            Dir::Right => self.x += 1,
        }
        match board[self.y][self.x] {
            '\\' => {
                self.dir = match self.dir {
                    Dir::Up => Dir::Left,
                    Dir::Down => Dir::Right,
                    Dir::Left => Dir::Up,
                    Dir::Right => Dir::Down,
                }
            }
            '/' => {
                self.dir = match self.dir {
                    Dir::Up => Dir::Right,
                    Dir::Down => Dir::Left,
                    Dir::Left => Dir::Down,
                    Dir::Right => Dir::Up,
                }
            }
            '+' => {
                self.dir = self.dir.turn(&self.turn);
                self.turn = self.turn.next();
            }
            _ => (),
        }
    }
}

impl Eq for Cart {}
impl PartialEq for Cart {
    fn eq(&self, other: &Self) -> bool {
        self.x == other.x && self.y == other.y
    }
}
impl Ord for Cart {
    fn cmp(&self, other: &Self) -> Ordering {
        match self.y.cmp(&other.y) {
            Ordering::Equal => self.x.cmp(&other.x),
            x => x,
        }
    }
}
impl PartialOrd for Cart {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

struct Sim {
    carts: Vec<Cart>,
    board: Vec<Vec<char>>,
}

impl Sim {
    fn new(carts: Vec<Cart>, board: Vec<Vec<char>>) -> Self {
        Self {
            carts: carts,
            board: board,
        }
    }

    fn tick(mut self) -> Self {
        let mut crashed = HashSet::new();
        for n in 0..self.carts.len() {
            if crashed.contains(&n) {
                continue;
            }
            {
                let (x, y) = &self.carts[n].next_step();
                if let Some(k) = self.detect_crash((*x, *y), &crashed) {
                    println!("{} {}", x, y);
                    crashed.insert(n);
                    crashed.insert(k);
                }
            }
            self.carts[n].tick(&self.board);
        }
        let mut next_carts = self
            .carts
            .into_iter()
            .enumerate()
            .filter(|(n, _)| !crashed.contains(n))
            .map(|(_, i)| i)
            .collect::<Vec<_>>();
        next_carts.sort();
        Self {
            carts: next_carts,
            board: self.board,
        }
    }

    fn detect_crash(&self, (x, y): (usize, usize), crashed: &HashSet<usize>) -> Option<usize> {
        for (n, i) in self.carts.iter().enumerate() {
            if crashed.contains(&n) {
                continue;
            }
            if i.x == x && i.y == y {
                return Some(n);
            }
        }
        None
    }
}

fn main() {
    let file = File::open(PUZZLEINPUT).expect("Failed to open file");
    let reader = BufReader::new(file);

    let mut sim = {
        let mut carts = Vec::new();
        let mut board = Vec::new();
        for line in reader.lines() {
            board.push(
                line.expect("Failed to read line")
                    .chars()
                    .collect::<Vec<_>>(),
            );
        }
        for (y, row) in board.iter_mut().enumerate() {
            for (x, i) in row.iter_mut().enumerate() {
                if let Some(k) = Dir::cart_char_to_dir(*i) {
                    match k {
                        Dir::Up => *i = '|',
                        Dir::Down => *i = '|',
                        Dir::Left => *i = '-',
                        Dir::Right => *i = '-',
                    }
                    carts.push(Cart::new(x, y, k));
                }
            }
        }
        Sim::new(carts, board)
    };

    while sim.carts.len() > 1 {
        sim = sim.tick();
    }

    let last_cart = &sim.carts[0];
    println!("{} {}", last_cart.x, last_cart.y);
}
